### PR TITLE
task 3. Implementation of the solution of the problem of Collatz sequence

### PR DESCRIPTION
**Pull request by rubikym**
#3 Rework task. Delete not needed classes. Now uses two classes (CollatzSequence, Program). CollatzSequence implement IEnumerable<ulong> and all calculation do in class implement IEnumerator<ulong>.

### CODE REVIEW

File: solutions/task-3/C#-Console/CollatzProject/CollatzProject/CollatzСalculations.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        public void Dispose() { }
+
+        public bool MoveNext()
+        {
+            if (!(currentValue < ((UInt64.MaxValue - 1) / 3)))
+            {
+                return false;
+            }
+
+            if (currentValue % 2 == 0)
```
[the remark this answers is not in the document]
Comment (rubikym) [reply solutions/task-3/C#-Console/CollatzProject/CollatzProject/CollatzСalculations.cs]: corrected

File: solutions/task-3/C#-Console/CollatzProject/CollatzProject/Program.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+using System.Threading.Tasks;
+
+namespace CollatzProject
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            var outputNumber = GetNaturalNumber();
+            Console.WriteLine(String.Format("Total steps completed:{0}", GetNumberOfSteps(outputNumber)));
```
[the remark this answers is not in the document]
Comment (rubikym) [reply solutions/task-3/C#-Console/CollatzProject/CollatzProject/Program.cs]: corrected

File: solutions/task-3/C#-Console/CollatzProject/CollatzProject/Program.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        private static ulong GetNaturalNumber()
+        {
+            var thisIsNumber = false;
+            var outputNumber = 0UL;
+            do
+            {
+                Console.WriteLine("Enter natural number");
+                var outputString = Console.ReadLine();
```
[the remark this answers is not in the document]
Comment (rubikym) [reply solutions/task-3/C#-Console/CollatzProject/CollatzProject/Program.cs]: corrected

File: solutions/task-3/C#-Console/CollatzProject/CollatzProject/Program.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        {
+            var outputNumber = GetNaturalNumber();
+            Console.WriteLine(String.Format("Total steps completed:{0}", GetNumberOfSteps(outputNumber)));
+            Console.ReadLine();
+        }
+
+        private static ulong GetNaturalNumber()
+        {
+            var thisIsNumber = false;
+            var outputNumber = 0UL;
```
[the remark this answers is not in the document]
Comment (rubikym) [reply solutions/task-3/C#-Console/CollatzProject/CollatzProject/Program.cs]: corrected

File: solutions/task-3/C#-Console/CollatzProject/CollatzProject/Program.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        static void Main(string[] args)
+        {
+            var outputNumber = GetNaturalNumber();
+            Console.WriteLine(String.Format("Total steps completed:{0}", GetNumberOfSteps(outputNumber)));
+            Console.ReadLine();
+        }
+
+        private static ulong GetNaturalNumber()
+        {
+            var thisIsNumber = false;
```
[the remark this answers is not in the document]
Comment (rubikym) [reply solutions/task-3/C#-Console/CollatzProject/CollatzProject/Program.cs]: corrected

File: solutions/task-3/C#-Console/CollatzProject/CollatzProject/Program.cs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+                case var _ when Int32.TryParse(outputString, out var i) && i <= 0:
+                    return "You can't enter value less than zero";
+                case var _ when Double.TryParse(outputString, out var _):
+                    return "You can't enter a fractional value";
+                default:
+                    return "you can not enter not number";
+            }
+        }
+
+        private static long GetNumberOfSteps(ulong outputNumber)
```
[the remark this answers is not in the document]
Comment (rubikym) [reply solutions/task-3/C#-Console/CollatzProject/CollatzProject/Program.cs]: corrected

File: solutions/task-3/C#-Console/CollatzProject/CollatzProject/Program.cs [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+            while (!thisIsNumber);
+            return outputNumber;
+        }
+
+        private static string GetErrorMessage(string outputString)
+        {
+            switch (outputString)
+            {
+                case var _ when Int32.TryParse(outputString, out var i) && i <= 0:
+                    return "You can't enter value less than zero";
```
[the remark this answers is not in the document]
Comment (rubikym) [reply solutions/task-3/C#-Console/CollatzProject/CollatzProject/Program.cs]: corrected

File: solutions/task-3/C#-Console/CollatzProject/CollatzProject/CollatzСalculations.cs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+                currentValue = currentValue / 2;
+            }
+            else
+            {
+                currentValue = currentValue * 3 + 1;
+            }
+            return true;
+        }
+
+        public void Reset() => currentValue = 1;
```
[the remark this answers is not in the document]
Comment (rubikym) [reply solutions/task-3/C#-Console/CollatzProject/CollatzProject/CollatzСalculations.cs]: corrected

File: solutions/task-3/C#-Console/CollatzProject/CollatzProject/CollatzСalculations.cs [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+        object IEnumerator.Current { get { return this.Current; } }
+
+        public void Dispose() { }
+
+        public bool MoveNext()
+        {
+            if (!(currentValue < ((UInt64.MaxValue - 1) / 3)))
+            {
+                return false;
```
[the remark this answers is not in the document]
Comment (rubikym) [reply solutions/task-3/C#-Console/CollatzProject/CollatzProject/CollatzСalculations.cs]: Now this metod trow exeption

File: solutions/task-3/C#-Console/CollatzProject/CollatzProject/Program.cs [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+            }
+            while (!thisIsNumber);
+            return outputNumber;
+        }
+
+        private static string GetErrorMessage(string outputString)
+        {
+            switch (outputString)
+            {
+                case var _ when Int32.TryParse(outputString, out var i) && i <= 0:
```
[the remark this answers is not in the document]
Comment (rubikym) [reply solutions/task-3/C#-Console/CollatzProject/CollatzProject/Program.cs]: This metod was deleted

File: solutions/task-3/C#-Console/CollatzProject/CollatzProject/Program.cs [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace CollatzProject
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            var outputNumber = GetNaturalNumber();
```
[the remark this answers is not in the document]
Comment (quixoticaxis) [reply solutions/task-3/C#-Console/CollatzProject/CollatzProject/Program.cs]: It was not fixed!

File: solutions/task-3/C#-Console/CollatzProject/CollatzProject/CollatzСalculations.cs [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+            
+            return true;
+        }
+
+        public void Reset() => currentValue = startingValue;
+    }
+
+    public class CollatzOverflowExeption : Exception
+    {
+        public override string Message => "Entered number caused overflow. Can't be calculated collatz sequence.";
```
Comment (quixoticaxis) [inline solutions/task-3/C#-Console/CollatzProject/CollatzProject/CollatzСalculations.cs]: "Cannot calculate the number of steps through Collatz sequence"

Comment (quixoticaxis) [reply solutions/task-3/C#-Console/CollatzProject/CollatzProject/CollatzСalculations.cs]: You should not have overriden this property. You could have set its value by calling an appropriate `base` constructor:
public CollatzOverflowExeption() : base("Entered number caused overflow. Cannot calculate the number of steps through Collatz sequence.") { }

**Pull request closed without merging**